File: src/conventions/core/pipeline/meaning-arbitrator.ts
import type {
  ArbitrationResult,
  EncodedProposal,
  EliminationRecord,
  PipelineCarrier,
  PipelineResult,
} from "../../../core/contracts/module-surface";
import type { Call } from "../../../engine/types";
import type { BidMeaning } from "../../../core/contracts/meaning";
import type {
  DecisionProvenance,
  EliminationTrace,
  ArbitrationTrace,
  LegalityTrace,
  EncodingTrace,
  ApplicabilityEvidence,
  HandoffTrace,
} from "../../../core/contracts/provenance";
import type { MeaningProposal } from "../../../core/contracts/meaning";
import { compareRanking, BAND_PRIORITY } from "../../../core/contracts/meaning";
import { evaluateProposal, classifyIntoSets } from "./arbitration-helpers";
import type {
  EvidenceBundle,
  ConditionEvidence,
  RejectionEvidence,
  AlternativeEvidence,
} from "../../../core/contracts/evidence-bundle";
import type { DeclaredEncoderKind } from "../../../core/contracts/agreement-module";
import type { EncoderConfig } from "./encoder-resolver";
import { formatCallForEvidence } from "../../../engine/call-helpers";

/** Input to the arbitrator: a proposal paired with the surface it came from. */
export interface ArbitrationInput {
  readonly proposal: MeaningProposal;
  readonly surface: {
    readonly encoding: { readonly defaultCall: Call };
    readonly encoderKind?: DeclaredEncoderKind;
    readonly encoderConfig?: EncoderConfig;
  };
}

/** Build ArbitrationInput[] by zipping proposals with their source surfaces. */
export function zipProposalsWithSurfaces(
  proposals: readonly MeaningProposal[],
  surfaces: readonly BidMeaning[],
): ArbitrationInput[] {
  return proposals.map((proposal, i) => {
    const surface = surfaces[i];
    if (!surface) {
      throw new Error(`zipProposalsWithSurfaces: no surface at index ${i}`);
    }
    return { proposal, surface };
  });
}

/**
 * Build a resolution function from semantic class alias pairs.
 * Each alias maps a module-local semanticClassId to a canonical ID.
 * When two proposals resolve to the same canonical ID, they are treated
 * as equivalent for deduplication (higher-ranked wins).
 */
function buildAliasResolver(
  aliases: readonly { from: string; to: string }[],
): (semanticClassId: string | undefined) => string | undefined {
  if (aliases.length === 0) {
    return (id) => id;
  }
  const aliasMap = new Map<string, string>();
  for (const { from, to } of aliases) {
    aliasMap.set(from, to);
  }
  return (id) => {
    if (id === undefined) return undefined;
    return aliasMap.get(id) ?? id;
  };
}

/**
 * Deduplicate a truth set by resolved semantic class.
 * For each group of carriers sharing the same resolved semanticClassId,
 * only the highest-ranked (first in sorted order) survives. Others are
 * eliminated with reason "semantic-class-alias-dedup".
 *
 * Carriers without a semanticClassId are never deduplicated.
 * The input array must already be sorted by ranking (highest first).
 */
function deduplicateBySemanticClassAlias(
  sorted: readonly PipelineCarrier[],
  resolveAlias: (id: string | undefined) => string | undefined,
): {
  deduplicated: readonly PipelineCarrier[];
  aliasEliminated: readonly PipelineCarrier[];
} {
  const seen = new Map<string, PipelineCarrier>();
  const deduplicated: PipelineCarrier[] = [];
  const aliasEliminated: PipelineCarrier[] = [];

  for (const carrier of sorted) {
    const resolved = resolveAlias(carrier.proposal.semanticClassId);

    // Carriers without a semanticClassId cannot be alias-deduplicated
    if (resolved === undefined) {
      deduplicated.push(carrier);
      continue;
    }

    const existing = seen.get(resolved);
    if (existing === undefined) {
      seen.set(resolved, carrier);
      deduplicated.push(carrier);
    } else {
      // This carrier is a lower-ranked duplicate — mark as eliminated
      const eliminatedCarrier: PipelineCarrier = {
        ...carrier,
        traces: {
          ...carrier.traces,
          elimination: {
            candidateId: carrier.proposal.meaningId,
            stage: "arbitration",
            reason: `Semantic class alias dedup: resolved to "${resolved}", already represented by "${existing.proposal.meaningId}"`,
            evidence: [],
            strength: "entailed",
          },
        },
      };
      aliasEliminated.push(eliminatedCarrier);
    }
  }

  return { deduplicated, aliasEliminated };
}

// ─── Provenance Helpers ────────────────────────────────────────

/** Map a MeaningClause to a ConditionEvidence (shared across provenance builders). */
function clauseToEvidence(c: { factId: string; satisfied: boolean; observedValue?: unknown; value: unknown }): ConditionEvidence {
  return { conditionId: c.factId, factId: c.factId, satisfied: c.satisfied, observedValue: c.observedValue, threshold: c.value };
}


/** Build EvidenceBundle from carriers. */
function buildEvidenceBundleFromCarriers(
  inputs: readonly ArbitrationInput[],
  eliminated: readonly PipelineCarrier[],
  finalTruthSet: readonly PipelineCarrier[],
  selected: PipelineCarrier | null,
): EvidenceBundle {
  const matched: EvidenceBundle["matched"] = selected
    ? { meaningId: selected.proposal.meaningId, satisfiedConditions: selected.proposal.clauses.map(clauseToEvidence) }
    : null;

  const rejected: RejectionEvidence[] = eliminated.map((c) => {
    const failedClauses = c.proposal.clauses.filter((cl) => !cl.satisfied);
    return {
      meaningId: c.proposal.meaningId,
      failedConditions: failedClauses.map(clauseToEvidence),
      negatableFailures: failedClauses.map((cl) => clauseToEvidence({ ...cl, satisfied: false })),
      moduleId: c.proposal.moduleId,
    };
  });

  const selectedFacts = selected ? new Set(selected.proposal.clauses.map(c => c.factId)) : new Set<string>();
  const alternatives: AlternativeEvidence[] = finalTruthSet
    .filter((c) => c !== selected)
    .map((c) => ({
      meaningId: c.proposal.meaningId,
      call: formatCallForEvidence(c.call),
      ranking: { band: c.proposal.ranking.recommendationBand, specificity: c.proposal.ranking.specificity },
      reason: "truth-set-member-not-selected",
      conditionDelta: c.proposal.clauses.filter(cl => !selectedFacts.has(cl.factId)).map(clauseToEvidence),
    }));

  return { matched, rejected, alternatives, exhaustive: true, fallbackReached: selected === null };
}

// ─── Main Arbitration ──────────────────────────────────────────

export function arbitrateMeanings(
  inputs: readonly ArbitrationInput[],
  options?: {
    legalCalls?: readonly Call[];
    semanticClassAliases?: readonly { from: string; to: string }[];
    handoffs?: readonly HandoffTrace[];
  },
): PipelineResult {
  // Step 1: Evaluate each proposal through gates (semantic, legality, encoding)
  const passedCarriers: PipelineCarrier[] = [];
  const eliminatedCarriers: PipelineCarrier[] = [];

  for (const input of inputs) {
    const result = evaluateProposal(input, options?.legalCalls);
    if (result.passedAllGates || result.addToAcceptable) {
      passedCarriers.push(result.carrier);
    }
    if (!result.passedAllGates) {
      eliminatedCarriers.push(result.carrier);
    }
  }

  // Step 2: Classify into truth set (all satisfied + legal) and acceptable set
  const { truthSet, acceptableSet } = classifyIntoSets(passedCarriers);

  // Step 3: Sort truth set by ranking, then deduplicate by semantic class alias
  const sortedTruth = [...truthSet].sort((a, b) =>
    compareRanking(a.proposal.ranking, b.proposal.ranking),
  );
  const resolveAlias = buildAliasResolver(options?.semanticClassAliases ?? []);
  const { deduplicated, aliasEliminated } =
    deduplicateBySemanticClassAlias(sortedTruth, resolveAlias);
  eliminatedCarriers.push(...aliasEliminated);

  // Step 4: Select winner (highest-ranked in deduplicated truth set)
  const finalTruthSet = deduplicated;
  const recommended = deduplicated;
  const selected: PipelineCarrier | null = recommended[0] ?? null;

  // Step 5: Build cross-surface provenance
  const allCarriers = [...passedCarriers, ...eliminatedCarriers.filter(c => !passedCarriers.includes(c))];
  const arbitrationTraces: ArbitrationTrace[] = allCarriers.map((c) => {
    const isTruth = finalTruthSet.includes(c) && c.eligibility.hand.satisfied && c.eligibility.encoding.legal;
    const recIndex = recommended.indexOf(c);
    return {
      candidateId: c.proposal.meaningId,
      truthSetMember: isTruth,
      acceptableSetMember: !c.eligibility.hand.satisfied && c.eligibility.encoding.legal,
      recommendationRank: recIndex >= 0 ? recIndex : undefined,
      rankingInputs: {
        recommendationBand: BAND_PRIORITY[c.proposal.ranking.recommendationBand],
        handFitScore: undefined,
        modulePrecedence: c.proposal.ranking.modulePrecedence ?? 0,
        specificity: c.proposal.ranking.specificity,
      },
    };
  });

  const applicability: ApplicabilityEvidence = selected
    ? {
      factDependencies: selected.proposal.clauses.map((c) => c.factId),
      evaluatedConditions: selected.proposal.clauses.map(clauseToEvidence),
    }
    : { factDependencies: [], evaluatedConditions: [] };

  const evidenceBundle = buildEvidenceBundleFromCarriers(inputs, eliminatedCarriers, finalTruthSet, selected);

  return {
    selected,
    truthSet: finalTruthSet,
    acceptableSet,
    recommended,
    eliminated: eliminatedCarriers,
    applicability,
    activation: [], // activation traces not yet implemented — always empty
    arbitration: arbitrationTraces,
    handoffs: options?.handoffs ?? [],
    evidenceBundle,
  };
}

// ─── Legacy Conversion Shims (temporary — removed in Phase 4) ──

/** Convert PipelineResult to ArbitrationResult for unmigrated consumers. */
export function pipelineResultToArbitration(result: PipelineResult): ArbitrationResult {
  const carrierToEncoded = (c: PipelineCarrier): EncodedProposal => ({
    proposal: c.proposal,
    call: c.call,
    isDefaultEncoding: c.isDefaultEncoding,
    legal: c.legal,
    allEncodings: c.allEncodings,
    eligibility: c.eligibility,
  });

  return {
    selected: result.selected ? carrierToEncoded(result.selected) : null,
    truthSet: result.truthSet.map(carrierToEncoded),
    acceptableSet: result.acceptableSet.map(carrierToEncoded),
    recommended: result.recommended.map(carrierToEncoded),
    eliminations: result.eliminated.map((c) => ({
      candidateBidName: c.proposal.meaningId,
      moduleId: c.proposal.moduleId,
      reason: c.traces.elimination?.reason ?? "Gate check failed",
      gateId: undefined,
    })),
    evidenceBundle: result.evidenceBundle,
  };
}

/** Convert PipelineResult to DecisionProvenance for unmigrated consumers. */
export function pipelineResultToProvenance(result: PipelineResult): DecisionProvenance {
  const allCarriers = [
    ...result.truthSet,
    ...result.acceptableSet,
    ...result.eliminated,
  ];
  return {
    applicability: result.applicability,
    activation: result.activation,
    encoding: allCarriers.map((c) => c.traces.encoding),
    legality: allCarriers.map((c) => c.traces.legality),
    arbitration: result.arbitration,
    eliminations: result.eliminated
      .filter((c) => c.traces.elimination !== undefined)
      .map((c) => c.traces.elimination!),
    handoffs: result.handoffs,
  };
}

File: crates/bridge-conventions/src/pipeline/evaluation/meaning_arbitrator.rs
//! Meaning arbitrator — tiered selection producing PipelineResult.
//!
//! Mirrors TS from `pipeline/evaluation/meaning-arbitrator.ts`.

use std::collections::{BTreeMap, HashSet};

use bridge_engine::types::Call;
use serde::Serialize;

use crate::pipeline::evaluation::arbitration_helpers::{classify_into_sets, evaluate_proposal};
use crate::pipeline::evaluation::provenance::{
    ApplicabilityEvidence, ArbitrationTrace, HandoffTrace,
};
use crate::pipeline::evaluation::types::{compare_ranking, MeaningClause, MeaningProposal};
use crate::pipeline::evidence_bundle::{
    AlternativeEvidence, AlternativeRanking, ConditionEvidence, EvidenceBundle, MatchedEvidence,
    RejectionEvidence,
};
use crate::pipeline::types::{PipelineCarrier, PipelineResult};

/// Arbitrate meanings — main entry point.
///
/// Gates → classify → sort → deduplicate → select.
pub fn arbitrate_meanings(
    proposals: &[MeaningProposal],
    is_legal: &dyn Fn(&Call) -> bool,
) -> PipelineResult {
    if proposals.is_empty() {
        return PipelineResult::empty();
    }

    // Step 1: Evaluate all proposals through gates
    let carriers: Vec<PipelineCarrier> = proposals
        .iter()
        .map(|p| evaluate_proposal(p, is_legal))
        .collect();

    // Step 2: Classify into truth and acceptable sets
    let (mut truth_set, acceptable_set) = classify_into_sets(&carriers);

    // Step 3: Sort truth set by ranking
    truth_set
        .sort_by(|a, b| compare_ranking(&a.encoded.proposal.ranking, &b.encoded.proposal.ranking));

    // Step 4: Deduplicate by semantic class, then by content fingerprint.
    // Semantic class catches author-declared equivalences; content dedup
    // catches identical surfaces from different modules that happen to
    // share the same meaning (e.g., Jacoby Transfers and Four-Way Transfers
    // both defining 2D→H).
    let deduped = deduplicate_by_semantic_class(&truth_set);
    let deduped = deduplicate_by_content(&deduped);

    // Step 5: Select winner (first in sorted, deduplicated truth set)
    let selected = deduped.first().cloned();

    // Step 6: Identify eliminated carriers
    let truth_ids: HashSet<&str> = truth_set
        .iter()
        .map(|c| c.encoded.proposal.meaning_id.as_str())
        .collect();
    let eliminated: Vec<PipelineCarrier> = carriers
        .iter()
        .filter(|c| !truth_ids.contains(c.encoded.proposal.meaning_id.as_str()))
        .cloned()
        .collect();

    // Build arbitration traces
    let arbitration: Vec<ArbitrationTrace> = carriers
        .iter()
        .map(|c| {
            let outcome = if selected.as_ref().map(|s| &s.encoded.proposal.meaning_id)
                == Some(&c.encoded.proposal.meaning_id)
            {
                "selected"
            } else if truth_ids.contains(c.encoded.proposal.meaning_id.as_str()) {
                "truth-set"
            } else {
                "eliminated"
            };
            ArbitrationTrace {
                meaning_id: c.encoded.proposal.meaning_id.clone(),
                module_id: c.encoded.proposal.module_id.clone(),
                outcome: outcome.into(),
                reason: c.traces.elimination.as_ref().map(|e| e.reason.clone()),
            }
        })
        .collect();

    // Build handoff traces from selected carrier's negotiation delta
    let handoffs: Vec<HandoffTrace> = Vec::new(); // Populated by adapter layer

    let matched_count = truth_set.len();
    let eliminated_count = eliminated.len();

    // Build evidence bundle from carriers
    let evidence_bundle = build_evidence_bundle(&selected, &truth_set, &eliminated);

    PipelineResult {
        selected,
        truth_set,
        acceptable_set,
        recommended: deduped,
        eliminated,
        applicability: ApplicabilityEvidence {
            total_surfaces: proposals.len(),
            matched_count,
            eliminated_count,
        },
        activation: Vec::new(), // Populated by observation layer
        arbitration,
        handoffs,
        evidence_bundle: Some(evidence_bundle),
    }
}

/// Build an EvidenceBundle from the arbitration results.
fn build_evidence_bundle(
    selected: &Option<PipelineCarrier>,
    truth_set: &[PipelineCarrier],
    eliminated: &[PipelineCarrier],
) -> EvidenceBundle {
    // Matched: selected carrier's satisfied conditions
    let matched = selected.as_ref().map(|carrier| {
        let conditions = carrier
            .proposal()
            .clauses
            .iter()
            .filter(|c| c.satisfied)
            .map(|c| clause_to_condition_evidence(c, true))
            .collect();
        MatchedEvidence {
            meaning_id: carrier.proposal().meaning_id.clone(),
            satisfied_conditions: conditions,
        }
    });

    // Rejected: eliminated carriers' failed conditions
    let rejected: Vec<RejectionEvidence> = eliminated
        .iter()
        .map(|carrier| {
            let failed = carrier
                .proposal()
                .clauses
                .iter()
                .filter(|c| !c.satisfied)
                .map(|c| clause_to_condition_evidence(c, false))
                .collect();
            RejectionEvidence {
                meaning_id: carrier.proposal().meaning_id.clone(),
                failed_conditions: failed,
                module_id: carrier.proposal().module_id.clone(),
                negatable_failures: None,
            }
        })
        .collect();

    // Alternatives: truth set entries that weren't selected
    let selected_id = selected.as_ref().map(|c| c.proposal().meaning_id.as_str());
    let alternatives: Vec<AlternativeEvidence> = truth_set
        .iter()
        .filter(|c| Some(c.proposal().meaning_id.as_str()) != selected_id)
        .map(|carrier| {
            let p = carrier.proposal();
            AlternativeEvidence {
                meaning_id: p.meaning_id.clone(),
                call: format!("{:?}", carrier.call()),
                ranking: AlternativeRanking {
                    band: format!("{:?}", p.ranking.recommendation_band),
                    specificity: p.ranking.specificity,
                },
                reason: "Alternative in truth set".to_string(),
                condition_delta: None,
            }
        })
        .collect();

    EvidenceBundle {
        matched,
        rejected,
        alternatives,
        exhaustive: true,
        fallback_reached: selected.is_none(),
    }
}

fn clause_to_condition_evidence(clause: &MeaningClause, satisfied: bool) -> ConditionEvidence {
    ConditionEvidence {
        condition_id: clause
            .clause_id
            .clone()
            .unwrap_or_else(|| clause.fact_id.clone()),
        fact_id: Some(clause.fact_id.clone()),
        satisfied,
        description: clause.description.clone(),
        observed_value: clause.observed_value.clone(),
        threshold: None,
        params: None,
    }
}

/// Deduplicate carriers by semantic class — keep highest-ranked per class.
///
/// NOTE: Only checks raw `semantic_class_id`. Alias resolution (cross-module
/// equivalence via aliased class IDs) is not implemented — spike (2026-04-02)
/// found no alias data in module fixtures and no cross-module duplicate class
/// IDs in any of the 4 bundles. Revisit if alias fields are added to fixtures.
fn deduplicate_by_semantic_class(sorted_carriers: &[PipelineCarrier]) -> Vec<PipelineCarrier> {
    let mut seen_classes: HashSet<String> = HashSet::new();
    let mut result = Vec::new();

    for carrier in sorted_carriers {
        let class_id = &carrier.encoded.proposal.semantic_class_id;
        if class_id.is_empty() || seen_classes.insert(class_id.clone()) {
            result.push(carrier.clone());
        }
    }

    result
}

/// Semantic fingerprint — fields that define meaning identity independent of
/// module provenance. Two surfaces with the same fingerprint are semantically
/// identical even if they come from different modules.
#[derive(Serialize)]
struct SemanticFingerprint<'a> {
    encoding: &'a crate::types::meaning::BidEncoding,
    clauses: Vec<ClauseFingerprint<'a>>,
    source_intent: IntentFingerprint<'a>,
    disclosure: crate::types::meaning::Disclosure,
}

#[derive(Serialize)]
struct ClauseFingerprint<'a> {
    fact_id: &'a str,
    operator: crate::types::meaning::FactOperator,
    value: &'a crate::types::meaning::ConstraintValue,
}

#[derive(Serialize)]
struct IntentFingerprint<'a> {
    intent_type: &'a str,
    /// BTreeMap ensures deterministic key ordering for canonical serialization.
    params: BTreeMap<&'a String, &'a serde_json::Value>,
}

/// Compute a canonical JSON string from the semantic fields of a proposal.
/// Identical surfaces from different modules produce the same key.
fn compute_semantic_key(proposal: &MeaningProposal) -> String {
    let fingerprint = SemanticFingerprint {
        encoding: &proposal.encoding,
        clauses: proposal
            .clauses
            .iter()
            .map(|c| ClauseFingerprint {
                fact_id: &c.fact_id,
                operator: c.operator,
                value: &c.value,
            })
            .collect(),
        source_intent: IntentFingerprint {
            intent_type: &proposal.source_intent.intent_type,
            params: proposal.source_intent.params.iter().collect(),
        },
        disclosure: proposal.disclosure,
    };
    serde_json::to_string(&fingerprint).expect("semantic fingerprint serialization cannot fail")
}

/// Deduplicate carriers by content — collapse surfaces with identical semantic
/// meaning (same encoding, clauses, source intent, disclosure) regardless of
/// which module contributed them. Keeps the highest-ranked carrier per content
/// fingerprint (input must be pre-sorted by ranking).
fn deduplicate_by_content(sorted_carriers: &[PipelineCarrier]) -> Vec<PipelineCarrier> {
    let mut seen: HashSet<String> = HashSet::new();
    let mut result = Vec::new();

    for carrier in sorted_carriers {
        let key = compute_semantic_key(&carrier.encoded.proposal);
        if seen.insert(key) {
            result.push(carrier.clone());
        }
    }

    result
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::adapter::tree_evaluation::{
        CandidateEligibility, EncodingEligibility, HandEligibility,
        PedagogicalEligibility,
    };
    use crate::pipeline::evaluation::provenance::{EncoderKind, EncodingTrace, LegalityTrace};
    use crate::pipeline::evaluation::types::{MeaningClause, RankingMetadata};
    use crate::pipeline::types::{CarrierTraces, EncodedProposal, EncodingOption};
    use crate::types::authored_text::{BidName, BidSummary, TeachingLabel};
    use crate::types::meaning::{
        BidEncoding, ConstraintValue, Disclosure, FactOperator, RecommendationBand, SourceIntent,
    };
    use bridge_engine::types::BidSuit;

    fn make_carrier_with_module(
        meaning_id: &str,
        module_id: &str,
        semantic_class_id: &str,
        call: Call,
        clauses: Vec<MeaningClause>,
        source_intent: SourceIntent,
        disclosure: Disclosure,
    ) -> PipelineCarrier {
        PipelineCarrier {
            encoded: EncodedProposal {
                proposal: MeaningProposal {
                    meaning_id: meaning_id.into(),
                    semantic_class_id: semantic_class_id.into(),
                    module_id: module_id.into(),
                    ranking: RankingMetadata {
                        recommendation_band: RecommendationBand::Must,
                        module_precedence: None,
                        declaration_order: 0,
                        specificity: 0.0,
                    },
                    clauses,
                    all_satisfied: true,
                    disclosure,
                    source_intent,
                    teaching_label: TeachingLabel {
                        name: BidName::new("test"),
                        summary: BidSummary::new("test"),
                    },
                    surface_bindings: None,
                    encoding: BidEncoding {
                        default_call: call.clone(),
                        alternate_encodings: None,
                    },
                    evidence: None,
                },
                call: call.clone(),
                is_default_encoding: true,
                legal: true,
                all_encodings: vec![EncodingOption {
                    call: call.clone(),
                    legal: true,
                }],
                eligibility: CandidateEligibility {
                    hand: HandEligibility {
                        satisfied: true,
                        failed_conditions: Vec::new(),
                    },
                    encoding: EncodingEligibility {
                        legal: true,
                        reason: None,
                    },
                    pedagogical: PedagogicalEligibility {
                        acceptable: true,
                        reasons: Vec::new(),
                    },
                },
            },
            traces: CarrierTraces {
                encoding: EncodingTrace {
                    encoder_kind: EncoderKind::DefaultCall,
                    considered_calls: None,
                    blocked_calls: None,
                },
                legality: LegalityTrace {
                    legal: true,
                    reason: None,
                },
                elimination: None,
            },
        }
    }

    fn transfer_intent() -> SourceIntent {
        SourceIntent {
            intent_type: "transfer".into(),
            params: std::collections::HashMap::from([(
                "target".into(),
                serde_json::Value::String("hearts".into()),
            )]),
        }
    }

    fn hcp_clause(min: i64) -> MeaningClause {
        MeaningClause {
            fact_id: "hand.hcp".into(),
            operator: FactOperator::Gte,
            value: ConstraintValue::int(min),
            satisfied: true,
            clause_id: None,
            description: None,
            observed_value: None,
            is_public: None,
        }
    }

    // ── deduplicate_by_semantic_class ─────────────────────────────────

    #[test]
    fn deduplicate_preserves_order() {
        let result = deduplicate_by_semantic_class(&[]);
        assert!(result.is_empty());
    }

    // ── deduplicate_by_content ───────────────────────────────────────

    #[test]
    fn content_dedup_collapses_identical_surfaces_from_different_modules() {
        let call = Call::Bid {
            level: 2,
            strain: BidSuit::Diamonds,
        };
        let clauses = vec![hcp_clause(0)];
        let intent = transfer_intent();

        let a = make_carrier_with_module(
            "jacoby:transfer-hearts",
            "jacoby-transfers",
            "",
            call.clone(),
            clauses.clone(),
            intent.clone(),
            Disclosure::Announcement,
        );
        let b = make_carrier_with_module(
            "fourway:transfer-hearts",
            "four-way-transfers",
            "",
            call,
            clauses,
            intent,
            Disclosure::Announcement,
        );

        let result = deduplicate_by_content(&[a, b]);
        assert_eq!(result.len(), 1, "identical content should collapse to one");
        assert_eq!(result[0].proposal().meaning_id, "jacoby:transfer-hearts");
    }

    #[test]
    fn content_dedup_preserves_different_surfaces() {
        let intent = transfer_intent();

        let a = make_carrier_with_module(
            "jacoby:transfer-hearts",
            "jacoby-transfers",
            "",
            Call::Bid {
                level: 2,
                strain: BidSuit::Diamonds,
            },
            vec![hcp_clause(0)],
            intent.clone(),
            Disclosure::Announcement,
        );
        let b = make_carrier_with_module(
            "fourway:transfer-clubs",
            "four-way-transfers",
            "",
            Call::Bid {
                level: 2,
                strain: BidSuit::Spades,
            },
            vec![hcp_clause(0)],
            intent,
            Disclosure::Alert,
        );

        let result = deduplicate_by_content(&[a, b]);
        assert_eq!(result.len(), 2, "different content should not collapse");
    }

    #[test]
    fn content_dedup_distinguishes_by_clauses() {
        let call = Call::Bid {
            level: 2,
            strain: BidSuit::Diamonds,
        };
        let intent = transfer_intent();

        let a = make_carrier_with_module(
            "mod-a:surface",
            "mod-a",
            "",
            call.clone(),
            vec![hcp_clause(0)],
            intent.clone(),
            Disclosure::Announcement,
        );
        let b = make_carrier_with_module(
            "mod-b:surface",
            "mod-b",
            "",
            call,
            vec![hcp_clause(6)],
            intent,
            Disclosure::Announcement,
        );

        let result = deduplicate_by_content(&[a, b]);
        assert_eq!(result.len(), 2, "different clause values should not collapse");
    }

    #[test]
    fn content_dedup_ignores_module_metadata() {
        let call = Call::Bid {
            level: 2,
            strain: BidSuit::Diamonds,
        };
        let clauses = vec![hcp_clause(0)];
        let intent = transfer_intent();

        // Same semantic content but different meaning_id, module_id, semantic_class_id, teaching_label
        let mut a = make_carrier_with_module(
            "mod-a:transfer",
            "module-alpha",
            "class-a",
            call.clone(),
            clauses.clone(),
            intent.clone(),
            Disclosure::Announcement,
        );
        a.encoded.proposal.teaching_label = TeachingLabel {
            name: BidName::new("Transfer (Jacoby)"),
            summary: BidSummary::new("Shows 5+ hearts"),
        };

        let mut b = make_carrier_with_module(
            "mod-b:transfer",
            "module-beta",
            "class-b",
            call,
            clauses,
            intent,
            Disclosure::Announcement,
        );
        b.encoded.proposal.teaching_label = TeachingLabel {
            name: BidName::new("Transfer (Four-Way)"),
            summary: BidSummary::new("Shows hearts"),
        };

        let result = deduplicate_by_content(&[a, b]);
        assert_eq!(
            result.len(),
            1,
            "different module metadata should not prevent dedup"
        );
    }

    // ── compute_semantic_key ─────────────────────────────────────────

    #[test]
    fn semantic_key_deterministic_for_identical_proposals() {
        let call = Call::Bid {
            level: 2,
            strain: BidSuit::Diamonds,
        };
        let clauses = vec![hcp_clause(0)];
        let intent = transfer_intent();

        let a = make_carrier_with_module(
            "a:x",
            "mod-a",
            "",
            call.clone(),
            clauses.clone(),
            intent.clone(),
            Disclosure::Announcement,
        );
        let b = make_carrier_with_module(
            "b:x",
            "mod-b",
            "",
            call,
            clauses,
            intent,
            Disclosure::Announcement,
        );

        let key_a = compute_semantic_key(&a.encoded.proposal);
        let key_b = compute_semantic_key(&b.encoded.proposal);
        assert_eq!(key_a, key_b);
    }
}
